scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/TaskAction.java

```diff
@@ -1,163 +1,189 @@
+        }
+        
+        private CompletableFuture<TaskAction> onException(Throwable ex) {
+            // Exception occured while reading the window, we will have to restart
+            // ReadNewWindowTaskAction - read a window from state defined in task.
+            long backoffTime = workerConfiguration.workerRetryBackoff.getRetryBackoffTimeMs(tryAttempt);
+            logger.atSevere().withCause(ex).log("Error while reading a CDC change. Task: %s. " +
+                    "Task state: %s. Will retry after backoff (%d ms).", task.id, task.state, backoffTime);
+            return delay(backoffTime)
+                    .thenApply(t -> new ReadNewWindowTaskAction(workerConfiguration, task, tryAttempt + 1));
+        }
+
+        @Override
+        public CompletableFuture<TaskAction> run() {
+            if (newState != null) {
+                workerConfiguration.transport.setState(task.id, newState);
+            }
+            try {
+                CompletableFuture<TaskAction> taskActionFuture = reader.nextChange().
+                        thenApply(change -> new ConsumeChangeTaskAction(workerConfiguration, task, reader, change, tryAttempt));
+                return FutureUtils.thenComposeExceptionally(taskActionFuture, this::onException);
+            } catch (Throwable ex) {
+                return onException(ex);
+            }
+        }
+    }
+
+    private static final class ConsumeChangeTaskAction extends TaskAction {
+        private final Reader reader;
+        private final Optional<RawChange> change;
+        private final int tryAttempt;
+
+        public ConsumeChangeTaskAction(WorkerConfiguration workerConfiguration, Task task, Reader reader, Optional<RawChange> change, int tryAttempt) {
+            super(workerConfiguration, task);
+            this.reader = Preconditions.checkNotNull(reader);
+            this.change = Preconditions.checkNotNull(change);
+            Preconditions.checkArgument(tryAttempt >= 0);
+            this.tryAttempt = tryAttempt;
+        }
+
+        private CompletableFuture<TaskAction> onException(Throwable ex) {
+            // Exception occured while consuming the change, we will have to restart
+            // ReadNewWindowTaskAction - read a window from state defined in task.
+            long backoffTime = workerConfiguration.workerRetryBackoff.getRetryBackoffTimeMs(tryAttempt);
+            logger.atSevere().withCause(ex).log("Error while executing consume() method provided to the library. Task: %s. " +
+                    "Task state: %s. Will retry after backoff (%d ms).", task.id, task.state, backoffTime);
+            return delay(backoffTime)
+                    .thenApply(t -> new ReadNewWindowTaskAction(workerConfiguration, task, tryAttempt + 1));
+        }
+        
+        @Override
+        public CompletableFuture<TaskAction> run() {
+            if (change.isPresent()) {
+                Task updatedTask = task.updateState(change.get().getId());

```

Comment:
The code doesn't work correctly when there are retries of reading a window:

For example, for such CDC data:
```
CREATE KEYSPACE ks WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};
CREATE TABLE ks.t(pk int, ck int, v int, PRIMARY KEY(pk, ck)) WITH cdc = {'enabled': true};

BEGIN BATCH 
INSERT INTO ks.t(pk, ck, v) VALUES (1, 1, 1); 
INSERT INTO ks.t(pk, ck, v) VALUES (1, 2, 2); 
INSERT INTO ks.t(pk, ck, v) VALUES (1, 3, 3); 
INSERT INTO ks.t(pk, ck, v) VALUES (1, 4, 4); 
INSERT INTO ks.t(pk, ck, v) VALUES (1, 5, 5); 
APPLY BATCH ; 

INSERT INTO ks.t(pk, ck, v) VALUES (1, 6, 6)
```

When there is an exception in `Driver3Reader::nextChange` during reading row `(1, 3, 3)`, a retry `ReadNewWindowTaskAction` is queued, but this queued `TaskAction` will not correctly start from beginning of batch, but it will skip the entire batch (start from `(1, 6, 6)`) - resulting in loss of data.

It seems that this line (`Task updatedTask = task.updateState(change.get().getId());`) somehow interferes with the logic.

(Below is my hacky way to inject an exception in `nextChange`; afterwards you can observe this behaviour in `Printer` app (with `--events`))
```
// Global
private static final AtomicInteger ai = new AtomicInteger(0);

        @Override
        public CompletableFuture<Optional<RawChange>> nextChange() {
            CompletableFuture<Optional<RawChange>> result = new CompletableFuture<>();
            findNext(result);
            return result.thenCompose(s -> {
                if (s.isPresent()) {
                    if (s.get().getCell("v").getInt() == 3) {
                        if (ai.getAndIncrement() == 0) {
                            CompletableFuture<Optional<RawChange>> injectedException = new CompletableFuture<>();
                            injectedException.completeExceptionally(new RuntimeException("Injected exception"));
                            return injectedException;
                        }
                    }
                }
                return CompletableFuture.completedFuture(s);
            });
        }
```

Reply:
A similar problem is also present in "non-Event" mode (not caused by this PR).

One thing that probably needs changing is `TaskState` - it only stores `lastConsumedChangeId` consisting of `cdc$stream_id` and `cdc$time`, NOT `cdc$batch_seq_no`. Note that Event consumer would never save `TaskState` in the middle of batch.

Reply:
Afaik, this behaviour is not changed in any way by this PR. I agree it is an issue, but I am not sure I am comfortable fixing it here. Or am I misunderstanding and somehow this aggravated by the changes here?